### PR TITLE
Blockinfo metadata

### PR DESCRIPTION
**Pull request by jmederosalvarado**
## Changes:
- Adds metadata to `BlockInfo` to tell whether the block is invalid or finalized.
- Makes changes to the `BlockInfoDecoder` to serialize the new metadata while being backwards compatible with the previous serialization scheme.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Core/BlockInfo.cs [hunk 1]
```diff
@@ -32,9 +39,18 @@ public BlockInfo(Keccak blockHash, in UInt256 totalDifficulty)
...
-        public bool IsFinalized { get; set; }
-        
+
+        public bool IsFinalized
+        {
+            get => FinalizationStatus == FinalizationStatus.Finalized;
+            set
+            {
+                FinalizationStatus = value ? FinalizationStatus.Finalized : FinalizationStatus.None;
+            }
```
Comment (xas) [inline src/Nethermind/Nethermind.Core/BlockInfo.cs:49]: If it's all depend on `FinalizationStatus`, I think `IsFinalized`should become a _getter_ only ?

File: src/Nethermind/Nethermind.Core/BlockInfo.cs [hunk 2]
```diff
@@ -19,6 +19,13 @@
 
 namespace Nethermind.Core
 {
+    public enum FinalizationStatus
+    {
+        None = 0,
+        Finalized = 1,
+        Invalid,
+    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/BlockInfo.cs:27]: 1. Should be [Flags], we want to use it more than just finalization.
2. Rename to BlockMetadata, its will only about finalization.
3. Use explicit values in the enum

File: src/Nethermind/Nethermind.Core/BlockInfo.cs [hunk 3]
```diff
@@ -32,9 +39,18 @@ public BlockInfo(Keccak blockHash, in UInt256 totalDifficulty)
...
-        
-        public bool IsFinalized { get; set; }
-        
+
+        public bool IsFinalized
+        {
+            get => FinalizationStatus == FinalizationStatus.Finalized;
+            set
+            {
+                FinalizationStatus = value ? FinalizationStatus.Finalized : FinalizationStatus.None;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/BlockInfo.cs:48]: Use flags logic

File: src/Nethermind/Nethermind.Core/BlockInfo.cs [hunk 4]
```diff
@@ -32,9 +39,18 @@ public BlockInfo(Keccak blockHash, in UInt256 totalDifficulty)
...
+            }
+        }
+
+        public FinalizationStatus FinalizationStatus { get; set; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/BlockInfo.cs:52]: Rename to Metadata

File: src/Nethermind/Nethermind.Serialization.Rlp/BlockInfoDecoder.cs [hunk 5]
```diff
@@ -92,7 +92,7 @@ public Rlp Encode(BlockInfo? item, RlpBehaviors rlpBehaviors = RlpBehaviors.None
             
             if (_chainWithFinalization)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Serialization.Rlp/BlockInfoDecoder.cs:93]: Drop _chainWithFinalization check, it will be for all chains

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Serialization.Rlp/BlockInfoDecoder.cs (hunk 5 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Serialization.Rlp/BlockInfoDecoder.cs:93]: or encode it if its != 0

File: src/Nethermind/Nethermind.Core.Test/Encoding/BlockInfoDecoderTests.cs [hunk 6]
```diff
@@ -60,5 +65,40 @@ private static void Roundtrip(bool valueDecode)
...
+            }
+
+            Rlp[] elements = new Rlp[chainWithFinalization ? 4 : 3];
+            elements[0] = Rlp.Encode(item.BlockHash);
+            elements[1] = Rlp.Encode(item.WasProcessed);
+            elements[2] = Rlp.Encode(item.TotalDifficulty);
+
+            if (chainWithFinalization)
+            {
+                elements[3] = Rlp.Encode(item.IsFinalized);
```
Comment (avalonche) [inline src/Nethermind/Nethermind.Core.Test/Encoding/BlockInfoDecoderTests.cs:98]: test encoding with multiple flags?

Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Core.Test/Encoding/BlockInfoDecoderTests.cs:102]: This method is for encoding in the way that it was done before added the metadata field, it is only used to test that the new decoder is backwards compatible, i.e. `BlockInfo` encoded before adding metadata can be decoded by the new decoder. However I added a few Metadata flags to the roundtrip test above.

File: src/Nethermind/Nethermind.Core/BlockInfo.cs (hunk 2 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Core/BlockInfo.cs]: Plugin-extensible could be messy. What if we switch from one plugin to another with different set of flags?

File: src/Nethermind/Nethermind.AuRa.Test/AuRaBlockFinalizationManagerTests.cs [hunk 7]
```diff
@@ -54,7 +54,7 @@ public void Initialize()
 
             _validatorStore.GetValidators(Arg.Any<long?>()).Returns(new Address[] {TestItem.AddressA, TestItem.AddressB, TestItem.AddressC});
             
-            Rlp.Decoders[typeof(BlockInfo)] = new BlockInfoDecoder(true);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AuRa.Test/AuRaBlockFinalizationManagerTests.cs:57]: We don't need to replace the default one now.

File: src/Nethermind/Nethermind.Consensus.AuRa/InitializationSteps/InitRlpAuRa.cs [hunk 8]
```diff
@@ -28,7 +28,7 @@ public class InitRlpAuRa : InitRlp
 
         public override Task Execute(CancellationToken cancellationToken)
         {
-            Rlp.Decoders[typeof(BlockInfo)] = new BlockInfoDecoder(true);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.AuRa/InitializationSteps/InitRlpAuRa.cs:31]: We don't need to replace the default one now.

**Review by LukaszRozmej — APPROVED** [review]
We can remove not needed AuRa decored overrides now.

Please don't merge until 1.12.7 is released.

File: src/Nethermind/Nethermind.Core/BlockInfo.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/BlockInfo.cs]: Yes there are potential conflicts on only 32 or 64 flags

File: src/Nethermind/Nethermind.AuRa.Test/AuRaBlockFinalizationManagerTests.cs (hunk 7 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.AuRa.Test/AuRaBlockFinalizationManagerTests.cs:57]: done

File: src/Nethermind/Nethermind.Consensus.AuRa/InitializationSteps/InitRlpAuRa.cs (hunk 8 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus.AuRa/InitializationSteps/InitRlpAuRa.cs:31]: no we can remove it

File: src/Nethermind/Nethermind.Core/BlockInfo.cs [hunk 9]
```diff
@@ -14,27 +14,52 @@
...
+                if (value)
+                {
+                    Metadata |= BlockMetadata.Finalized;
+                }
+                else
+                {
+                    Metadata &= ~BlockMetadata.Finalized;
+                }
+            }
+        }
```
Comment (xas) [inline src/Nethermind/Nethermind.Core/BlockInfo.cs:59]: Sorry but I don't understand this code. You use a setter to change only another property
The only IsFinalized.set is on the Consensus.Aura project. Could it not be updated ?
Plus I think it's a dangerous practice, as `Metadata` can have technically its value modified in two different place (not on the property itself + difficulty to debug)

This code should be enough :
```suggestion
        public bool IsFinalized => (Metadata & BlockMetadata.Finalized) == BlockMetadata.Finalized;
```

**Pull request merged**